modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotWithMetastorageTest.java

```diff
@@ -0,0 +1,221 @@
+
+            return null;
+        });
+
+        queue.offer(testTask);
+
+        assertTrue(GridTestUtils.waitForCondition(() -> queue.size() > 10, getTestTimeout()));
+
+        ignite.context().cache().context().exchange()
+            .registerExchangeAwareComponent(new PartitionsExchangeAware() {
+                /** {@inheritDoc} */
+                @Override public void onInitAfterTopologyLock(GridDhtPartitionsExchangeFuture fut) {
+                    latch.countDown();
+                }
+            });
+
+        ignite.snapshot().createSnapshot(SNAPSHOT_NAME).get();
+
+        stop.set(true);
+        updFut.get();
+
+        stopAllGrids();
+
+        IgniteEx snp0 = startGridsFromSnapshot(Collections.singleton(0),
+            cfg -> resolveSnapshotWorkDirectory(cfg).getAbsolutePath(), SNAPSHOT_NAME, false);
+
+        Set<String> allKeys = new HashSet<>();
+        snp0.context().distributedMetastorage().iterate(SNAPSHOT_PREFIX, (key, val) -> allKeys.add(key));
+
+        stopGrid(0);
+
+        IgniteEx snp1 = startGridsFromSnapshot(Collections.singleton(1),
+            cfg -> resolveSnapshotWorkDirectory(cfg).getAbsolutePath(), SNAPSHOT_NAME, false);
+
+        // Iterator reads keys from the node heap map.
+        snp1.context().distributedMetastorage()
+            .iterate(SNAPSHOT_PREFIX, new BiConsumer<String, Serializable>() {
+                @Override public void accept(String key, Serializable value) {
+                    try {
+                        assertTrue("Keys must be the same on all nodes [key=" + key + ", all=" + allKeys + ']',
+                            allKeys.contains(key));
+                    }
+                    catch (Throwable t) {
+                        fail("Exception reading metastorage: " + t.getMessage());
+                    }
+                }
+            });
+    }
+
+    /** @throws Exception If fails. */
+    @Test
+    public void testMetastorageUpdateOnSnapshotFail() throws Exception {
+        AtomicInteger keyCnt = new AtomicInteger();
+        AtomicBoolean stop = new AtomicBoolean();
+
+        IgniteEx ignite = startGridsWithCache(2, dfltCacheCfg, CACHE_KEYS_RANGE);
+
+        IgniteInternalFuture<?> updFut = GridTestUtils.runMultiThreadedAsync(() -> {
+            while (!Thread.currentThread().isInterrupted() && !stop.get()) {
+                try {
+                    ignite.context().distributedMetastorage()
+                        .write(SNAPSHOT_PREFIX + keyCnt.getAndIncrement(), "value");
+                }
+                catch (IgniteCheckedException e) {
+                    throw new IgniteException(e);
+                }
+            }
+        }, 3, "dms-updater");
+
+        GridCacheSharedContext<?, ?> cctx = ignite.context().cache().context();
+        GridCacheDatabaseSharedManager db = (GridCacheDatabaseSharedManager)cctx.database();
+
+        db.addCheckpointListener(new CheckpointListener() {
+            /** {@inheritDoc} */
```

Comment:
As far as I know, there is no point in documenting methods in anonymous classes as this document will be completely ignored